ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/LedgerBridge.scala

```diff
@@ -0,0 +1,128 @@
+
+package com.daml
+package ledger.sandbox.bridge
+
+import akka.NotUsed
+import akka.stream.scaladsl.Flow
+import com.codahale.metrics.InstrumentedExecutorService
+import com.daml.error.ErrorCodesVersionSwitcher
+import com.daml.ledger.offset.Offset
+import com.daml.ledger.participant.state.index.v2.IndexService
+import com.daml.ledger.participant.state.kvutils.app.{Config, ParticipantConfig}
+import com.daml.ledger.participant.state.v2.Update
+import com.daml.ledger.resources.ResourceOwner
+import com.daml.ledger.sandbox.BridgeConfig
+import com.daml.ledger.sandbox.domain.Submission
+import com.daml.lf.data.{Ref, Time}
+import com.daml.lf.transaction.CommittedTransaction
+import com.daml.logging.LoggingContext
+import com.daml.metrics.Metrics
+import com.daml.platform.server.api.validation.ErrorFactories
+import com.google.common.primitives.Longs
+
+import java.util.UUID
+import java.util.concurrent.Executors
+import scala.concurrent.ExecutionContext
+
+trait LedgerBridge {
+  def flow: Flow[Submission, (Offset, Update), NotUsed]
+}
+
+object LedgerBridge {
+  def owner(
+      config: Config[BridgeConfig],
+      participantConfig: ParticipantConfig,
+      indexService: IndexService,
+      metrics: Metrics,
+  )(implicit
+      loggingContext: LoggingContext
+  ): ResourceOwner[LedgerBridge] = {
+    val bridgeMetrics = new BridgeMetrics(metrics)
+
+    implicit val ec: ExecutionContext =
+      ExecutionContext.fromExecutorService(
+        new InstrumentedExecutorService(
+          Executors.newWorkStealingPool(config.extra.bridgeThreadPoolSize),
+          metrics.registry,
+          bridgeMetrics.threadpool.toString,
+        )
+      )
+
+    if (config.extra.conflictCheckingEnabled)
+      for {
+        initialLedgerEnd <- ResourceOwner.forFuture(() => indexService.currentLedgerEnd())
+        conflictCheckingLedgerBridge = new ConflictCheckingLedgerBridge(
+          participantId = participantConfig.participantId,
+          indexService = indexService,
+          initialLedgerEnd =
+            Offset.fromHexString(Ref.HexString.assertFromString(initialLedgerEnd.value)),
+          bridgeMetrics = bridgeMetrics,
+          errorFactories = ErrorFactories(
+            new ErrorCodesVersionSwitcher(config.enableSelfServiceErrorCodes)
+          ),
+        )
+      } yield conflictCheckingLedgerBridge
+    else
+      ResourceOwner.forValue(() =>
+        new PassThroughLedgerBridge(participantId = participantConfig.participantId)
+      )
+  }
+
+  private[bridge] def toOffset(index: Long): Offset = Offset.fromByteArray(Longs.toByteArray(index))
+
+  private[bridge] def fromOffset(offset: Offset): Long = {
+    val offsetBytes = offset.toByteArray
+    if (offsetBytes.length > 8)
+      throw new RuntimeException(s"Byte array too big: ${offsetBytes.length}")
+    else
+      Longs.fromByteArray(
+        Array.fill[Byte](8 - offsetBytes.length)(0) ++ offsetBytes
+      )
+  }
+
+  def successMapper(
```

Comment:
This would be easier to test as a pure function

Reply:
Will extract in the unit testing PR